cpg_pipes/pipeline/pipeline.py

```diff
@@ -0,0 +1,960 @@
+        
+        if additional_stages_dict:
+            logger.info(
+                f'Additional required stages added as "skipped": '
+                f'{additional_stages_dict.keys()}'
+            )
+            for name, stage_ in self._stages_dict.items():
+                additional_stages_dict[name] = stage_    
+            self._stages_dict = additional_stages_dict
+
+        logger.info(f'Setting stages: {", ".join(s.name for s in stages if not s.skipped)}')
+        required_skipped_stages = [s for s in stages if s.skipped and s.required]
+        if required_skipped_stages:
+            logger.info(
+                f'Skipped stages that are used to get inputs: '
+                f'{", ".join(s.name for s in required_skipped_stages)}'
+            )
+
+        # Second round - actually adding jobs from the stages.
+        for i, (_, stage_) in enumerate(self._stages_dict.items()):
+            if not stage_.skipped:
+                logger.info(f'*' * 60)
+                logger.info(f'Stage {stage_.name}')
+
+            if stage_.required:
+                logger.info(f'Adding jobs for stage {stage_.name}')
+                stage_.output_by_target = stage_.add_to_the_pipeline(self)
+
+            if not stage_.skipped:
+                logger.info(f'')
+                if last_stage_num and i >= last_stage_num:
+                    logger.info(f'Last stage is {stage_.name}, stopping here')
+                    break
+
+    def get_datasets(self, only_active: bool = True) -> list[Dataset]:
+        """
+        Thin wrapper around corresponding Cohort method.
+        """
+        return self.cohort.get_datasets(only_active=only_active)
+
+    def add_dataset(self, name: str) -> Dataset:
+        """
+        Thin wrapper around corresponding Cohort method.
+        """
+        return self.cohort.add_dataset(name)
+
+    def get_all_samples(self, only_active: bool = True) -> list[Sample]:
+        """
+        Thin wrapper around corresponding Cohort method.
+        """
+        return self.cohort.get_samples(only_active=only_active)
+
+    def get_all_sample_ids(self, only_active: bool = True) -> list[str]:
+        """
+        Thin wrapper around corresponding Cohort method.
+        """
+        return self.cohort.get_sample_ids(only_active=only_active)
+
+
+# Type variable to make sure a Stage subclass always matches the
+# correspondinng Target subclass
+TargetT = TypeVar('TargetT', bound=Target)
+
+ExpectedResultT = Union[Path, dict[str, Path], None]
+
+StageOutputData = Union[
+    Path, hb.Resource, dict[str, Path], dict[str, hb.Resource]
+]
+
+
+# noinspection PyShadowingNames
+class StageOutput:
+    """
+    Represents a result of a specific stage, which was run on a specific target.
+    Can be a file path, or a Hail Batch Resource. Optionally wrapped in a dict.
+    """
+    def __init__(
+        self,
+        data: StageOutputData,
+        stage: 'Stage',
```

Comment:
Does the StageOutput / Input need a definition to the stage it's in? I can see it being relevant, but function definitions below do pass in a stage.